commercial/app/controllers/commercial/Multi.scala

```diff
@@ -21,94 +24,95 @@ class Multi(bestsellersAgent: BestsellersAgent,
-            views.html.jobs.jobsBlended(_, clickMacro)
-          }
-        }
-      case ("books", isbn) if isbn.nonEmpty =>
-        bestsellersAgent.getSpecificBooks(Seq(isbn)) map { books =>
-          books.headOption orElse {
-            bestsellersAgent.bestsellersTargetedAt(segment).headOption
-          } map {
-            views.html.books.booksBlended(_, clickMacro)
-          }
-        }
-      case ("books", _) =>
-        Future.successful {
-          bestsellersAgent.bestsellersTargetedAt(segment).headOption map {
-            views.html.books.booksBlended(_, clickMacro)
-          }
-        }
-      case ("travel", travelId) if travelId.nonEmpty =>
-        Future.successful {
-          travelOffersAgent.specificTravelOffers(Seq(travelId)).headOption orElse {
-            travelOffersAgent.offersTargetedAt(segment).headOption
-          } map {
-            views.html.travel.travelBlended(_, clickMacro)
-          }
-        }
-      case ("travel", _) =>
-        Future.successful {
-          travelOffersAgent.offersTargetedAt(segment).headOption map {
-            views.html.travel.travelBlended(_, clickMacro)
-          }
-        }
-      case ("masterclasses", eventBriteId) if eventBriteId.nonEmpty =>
-        Future.successful {
-          masterclassAgent.specificMasterclasses(Seq(eventBriteId)).filterNot(_.mainPicture.isEmpty).headOption orElse {
-            masterclassAgent.masterclassesTargetedAt(segment).filterNot(_.mainPicture.isEmpty).headOption
-          } map {
-            views.html.masterclasses.masterclassesBlended(_, clickMacro)
-          }
-        }
-      case ("masterclasses", _) =>
-        Future.successful {
-          masterclassAgent.masterclassesTargetedAt(segment).filterNot(_.mainPicture.isEmpty).headOption map {
-            views.html.masterclasses.masterclassesBlended(_, clickMacro)
-          }
-        }
-      case ("soulmates", _) =>
-        Future.successful {
-          for {
-            woman <- SoulmatesAgent.womenAgent.sample().headOption
-            man <- SoulmatesAgent.menAgent.sample().headOption
-          } yield {
-            views.html.soulmates.soulmatesBlended(Random.shuffle(Seq(woman, man)), clickMacro)
-          }
-        }
-      case _ => Future.successful(None)
-    }
+    val eventualContents = multiSample(requestedContent, slotIds, segment)
+
+    eventualContents map { contents =>
+      val content = contents map {
+        case b: Book        => views.html.books.booksBlended(b, clickMacro)
+        case j: Job         => views.html.jobs.jobsBlended(j, clickMacro)
+        case m: Masterclass => views.html.masterclasses.masterclassesBlended(m, clickMacro)
+        case p: MemberPair  => views.html.soulmates.soulmatesBlended(Random.shuffle(Seq(p.member1, p.member2)), clickMacro)
+        case t: TravelOffer => views.html.travel.travelBlended(t, clickMacro)
+        case _              => Html("")
```

Comment:
`LiveEvent` will need to be added to the `ADT` for that to be exhaustive otherwise you will always need the `_`. You must not be matching on the `ADT` if this is compiling and there is a type outside of it.